### PR TITLE
Proxy

### PR DESCRIPTION
**Pull request by proteus1121**
https://trello.com/c/FMxCB0CE/6-proxy-3rd-party-charts-service

### CODE REVIEW

File: charts/src/main/java/chart/desk/controllers/ChartProxyController.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    public ChartProxyController(ChartService chartService, YamlParser yamlParser, RestTemplate restTemplate) {
+        this.chartService = chartService;
+        this.yamlParser = yamlParser;
+        this.restTemplate = restTemplate;
+    }
+
+    @PostMapping("/proxy/{userId}")
+    public Map<Boolean, Long> proxyIndex(@PathVariable("userId") String userId,
+            @RequestParam("thirdPartyUrl") String thirdPartyUrl) throws JsonProcessingException {
+        String chartIndex = restTemplate.getForObject(thirdPartyUrl, String.class);
```
Comment (leafchild0) [inline charts/src/main/java/chart/desk/controllers/ChartProxyController.java:51]: You have to construct URL object from it first or similar: https://stackoverflow.com/questions/2230676/how-to-check-for-a-valid-url-in-java. I would also vote to move this into body so it's harder to provide any malicious url

File: charts/src/main/java/chart/desk/controllers/ChartProxyController.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        ChartIndex index = yamlParser.download(chartIndex);
+        Map<String, List<ChartEntry>> existedCharts = chartService.getIndex(userId).getEntries();
+        try (PoolingHttpClientConnectionManager connectionManager = new PoolingHttpClientConnectionManager();
```
Comment (leafchild0) [inline charts/src/main/java/chart/desk/controllers/ChartProxyController.java:54]: Why do you need both http client and rest template?

File: charts/src/main/java/chart/desk/controllers/ChartProxyController.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    }).collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+        } catch (Exception e) {
+            log.error("Proxy helm repository {} failed.", thirdPartyUrl, e);
+            throw new RuntimeException("Proxy helm repository " + thirdPartyUrl + " failed.");
+        }
+    }
+
+    private byte[] downloadFirstChart(List<String> urls, CloseableHttpClient client) {
+        for (String url : urls) {
+            HttpGet request = new HttpGet(url);
```
Comment (leafchild0) [inline charts/src/main/java/chart/desk/controllers/ChartProxyController.java:83]: Same question: why not use either Spring tools or Java built-in tools for this?

File: charts/src/main/java/chart/desk/model/HelmProperties.java [hunk 4]
```diff
@@ -6,6 +6,10 @@
 import java.util.Arrays;
 import java.util.Optional;
 
+/**
+ * @deprecated replace with ChartEntity
```
Comment (leafchild0) [inline charts/src/main/java/chart/desk/model/HelmProperties.java:10]: Can you elaborate on this?

Comment (proteus1121) [reply charts/src/main/java/chart/desk/model/HelmProperties.java:10]: will be done in https://trello.com/c/s3IxcWde/4-save-providence-metadata

File: charts/src/main/java/chart/desk/controllers/ChartProxyController.java (hunk 2 above)
Comment (proteus1121) [reply charts/src/main/java/chart/desk/controllers/ChartProxyController.java:59]: I need to download a lot of charts in parallel, so using apache http client for it. rest template for fast load and parsing yaml in object

**Review by leafchild0 — APPROVED** [review]

**Pull request merged**